### ISSUE TITLE
[QUESTION] Combining proxies and their whitelist/blacklsit settings from two FoxyProxy generations

### ISSUE THREAD

**Issue by COOLak** [foxyproxy/browser-extension#127]
After the infamous update that led to seemingly destroying all people's settings (but then turned out to be recoverable), I decided to keep the old version of FoxyProxy on my Windows PC (7.5.2), but use the new one on my Mac (Firefox in both cases).

Everything is fine with both versions, except a lot of times I have to duplicate my whitelists for some proxies on both devices, because sync doesn't work between these two generations.

Is there a method I could use to at least manually combine all into one without duplicates? I would just do it from time to time, at least.

**Comment by erosman** [comment]
> seemingly destroying all people's settings 

Not all. The bug only affected those who had `sync` enabled, and some who had `null` value for their `log` setting.

> Is there a method I could use to at least manually combine all into one without duplicates? 

Do both version have the same settings?
If that is the case, you can:
- Do all changes on the v7
- Export v7 settings
- Import the setting into v8 via "Import from older versions"

**Comment by COOLak** [comment]
> Do both version have the same settings?

That's exactly my point, no, they don't. I occasionally add or remove some rules on V7 and on V8 independently. I was wondering if I could combine them into one from time to time.

**Comment by erosman** [comment]
In that case, it wont matter even if they were both the same version since sync would make them both identical, but you want to have different settings on different machines.

**Comment by COOLak** [comment]
@erosman 
> In that case, it wont matter even if they were both the same version since sync would make them both identical, but you want to have different settings on different machines.

No! I want exactly the same settings on both machines. In fact, I've already upgraded to to the latest version even on the Windows PC (I see it has improved). But even then, I keep having different, individual sets of settings on PC and Mac. Both are using Firefox, and sync in enabled in both Firefox and FoxyProxy. And for some reason I still have different settings on two machines. I want the to get combined. How do I achive this? MaybeI could provide you some logs to explore?

**Comment by erosman** [comment]
I see. in that case, they should sync properly.
Please note that the mode is not synced.

Try this:
- Back-up your settings (Export)
- Uninstal FoxyProxy on one machine 
- Restart Firefox (to make sure Firefox cache is cleared)
- Re-instal FoxyProxy on that machine
- Enable sync on that machine (both Firefox sync & FoxyProxy sync)
- SAVE
- On the OTHER machine (with data/proxies), make some changes (e.g. disable/enable one proxy)
- SAVE
- Wait to see if the data from that machine is passed to the machine with no data

**Comment by COOLak** [comment]
I think you're still missing the point. I have two different sets of settings on two machines. Even if I export them both, I can then import only one of them, I would have to do all the other changes (from one of the machines) manually. Which I want to avoid and just combine these two sets into one.

**Comment by erosman** [comment]
Therefore, is this the situation?

- You have 2 different settings
- You want to combine the 2 sets into one
- After combining, you want to have identical settings on both machines


If that is the case, you have to manually combine them, which is easy with just copy & paste.
Export settings from both, copy proxies under one `data` to the other one.

Sync is not designed to combine data.

**Comment by COOLak** [comment]
Ah, well, not only proxies, a lot of different rules for each proxy too. But if there's no other way that doing it manually, so be it (or I'll just keep everything as is).

**Comment by erosman** [comment]
> Ah, well, not only proxies, a lot of different rules for each proxy too.

They are all in the same place, within the proxies.

- Export settings from both versions
- Open them in a Text Editor
- The settings will have the following format
```
{
  "mode": "pattern",
  "sync": true,
  "autoBackup": true,
  "showPatternProxy": false,
  "passthrough": "",
  "container": {  },
  "commands": {  },
  "data": [
    { ... proxy 1 (with its patterns) ... },
    { ... proxy 2 (with its patterns) ... },
    { ... proxy 3 (with its patterns) ... },
    { ... proxy 4 (with its patterns) ... },
    { ... proxy 5 (with its patterns) ... }
  ]
}
```
- Copy what is inside `data` from one
- Paste inside `data` in the other
- Remember that there should be a `,` after each proxy, except the last one
- SAVE
- Import the edited settings

> But if there's no other way that doing it manually, so be it

If there is a popular demand, we can add a feature to do that.

**Comment by COOLak** [comment]
> If there is a popular demand, we can add a feature to do that.

That would be cool. The method you described via text editor is even more mundane than just plain comparing the sets of settings and adding everything that's missing on one of the devices from the other device.

**Comment by COOLak** [comment]
Hmmm, however, I have an obvious question. How exactly does your sync option work. I have sync enabled on both devices but they both have different sets of proxies as well as settings within those proxies (even within the same ones). If I were to sync this extension to some third device, from which of my devices will the settings get imported? I'm curious of how it is decided.

**Comment by erosman** [comment]
> That would be cool. The method you described via text editor is even more mundane than just plain comparing the sets of settings and adding everything that's missing on one of the devices from the other device and vice versa.

Even if a feature is added, it would only add one sets to the other. It is not feasible to compare the proxies, and check their patterns and the order of the patterns and remove duplicates.

> however, I have an obvious question. How exactly does your sync option work. 

- If FoxyProxy Sync options is enable, every time FoxyPrxoy's Options are edited, it sends all proxies to the browser's special Sync storage area
- Sending data to other devices is performed by the browser
- If Browser sync is enabled, browser periodically checks if the Sync storage area has data, and sends that data to the other device that have browser sync enabled
- Other devices then sends the data to the corresponding extensions
- The FoxyProxy in other devices gets notified of the new data (if FoxyProxy sync option is enabled)
- FoxyProxy then, clears proxies from the receiving device and replaces it with the data from the sending device


For example:
_(each proxy data is complete with its patterns)_

- Device X with proxies a, b, c
- Device Y with proxies d, e, f
- Device X proxies are edited (e.g. one proxy is disabled)
- Sync is sent to device Y
- Device Y removes d, e, f and replaces them with a, b, c
- Now both devices only have proxies a, b, c

**Comment by erosman** [comment]
> Well, it does NOT happen in my case. Can I help you troubleshoot it please? What should I send?

Did you try the process in the https://github.com/foxyproxy/browser-extension/issues/127#issuecomment-1976697218?
